FaunaDB.Client/Types/Api21/RecursiveRef.cs

```diff
@@ -0,0 +1,123 @@
+            );
+
+            writer.WriteObject("@ref", props);
+        }
+
+        public override bool Equals(Expr v)
+        {
+            if (v == null || GetType() != v.GetType())
+                return false;
+
+            var w = v as ObjectRefV;
+
+            if (!Id.Equals(w.Id))
+                return false;
+
+            if (Class != null && !Class.Equals(w.Class))
+                return false;
+
+            if (Database != null && !Database.Equals(w.Database))
+                return false;
+
+            return true;
+        }
+
+        protected override int HashCode() =>
+            new { Id, Class, Database }.GetHashCode();
+
+        public override string ToString() =>
+            $"{GetType().Name}(id: \"{Id}\", class: {Option.Of(Class)}, database: {Option.Of(Database)})";
+    }
+
+    public class DatabaseV : ObjectRefV
+    {
+        public DatabaseV(string id, DatabaseV database = null) : base(id, ClassV.BuiltIn.DATABASES, database)
+        { }
+    }
+
+    public class ClassV : ObjectRefV
+    {
+        internal ClassV(string builtin) : base(builtin, null, null)
+        { }
+
+        public ClassV(string id, DatabaseV database = null) : base(id, BuiltIn.CLASSES, database)
+        { }
+
+        public static class BuiltIn
+        {
+            public static readonly ClassV CLASSES = new ClassV(builtin: "classes");
+            public static readonly ClassV INDEXES = new ClassV(builtin: "indexes");
+            public static readonly ClassV DATABASES = new ClassV(builtin: "databases");
+            public static readonly ClassV KEYS = new ClassV(builtin: "keys");
+            public static readonly ClassV FUNCTIONS = new ClassV(builtin: "functions");
+
+            internal static ClassV FromName(string name)
+            {
+                switch (name)
+                {
+                    case "classes": return CLASSES;
+                    case "indexes": return INDEXES;
+                    case "databases": return DATABASES;
+                    case "keys": return KEYS;
+                    case "functions": return FUNCTIONS;
+                }
+
+                throw new NotSupportedException($"The builtin class '{name}' is not supported");
+            }
+        }
+    }
+
+    public class IndexV : ObjectRefV
+    {
+        public IndexV(string id, DatabaseV database = null) : base(id, ClassV.BuiltIn.INDEXES, database)
+        { }
+    }
+
+    public class FunctionV : ObjectRefV
+    {
+        public FunctionV(string id, DatabaseV database = null) : base(id, ClassV.BuiltIn.FUNCTIONS, database)
+        { }
+    }
+
+    public class KeyV : ObjectRefV
+    {
+        public KeyV(string id, DatabaseV database = null) : base(id, ClassV.BuiltIn.KEYS, database)
+        { }
+    }
+
+    public class SelfV : ObjectRefV
+    {
+        public SelfV(ClassV @class) : base("self", @class, null)
+        { }
+    }
+
+    public class InstanceV : ObjectRefV
```

Comment:
I understand that once we enable 2.1 it will be the default ref, but this is still optional. The old ref need to keep working and also the new and old have different shapes: `{"@ref": "indexes/all_widgets"}` => `{"@ref": {"id": "all_widgets", "class": {"@ref": "classes"}}}`